lib/functions/clocksource.functions.sh

```diff
@@ -0,0 +1,147 @@
+
+_clocksource_write_config() {
+    local default="$1" clocksource="$2"
+    shift 2
+
+    cat <<EOF > $cw_CLOCKSOURCE_config_file
+################################################################################
+##
+## Alces Clusterware - Clocksource configuration
+## Copyright (c) 2017 Alces Software Ltd
+##
+################################################################################
+cw_CLUSTER_default_clocksource=${default}
+cw_CLUSTER_clocksource=${clocksource}
+EOF
+}
+
+_clocksource_get_available() {
+    cat /sys/devices/system/clocksource/clocksource0/available_clocksource
+}
+
+_clocksource_validate_source() {
+    local input="$1"
+    shift
+
+    validated=$(_clocksource_get_available | grep -wo "$input")
+    if [[ -z $validated ]]; then
+        action_die "unrecognised clocksource: $input"
+    fi
+}
+
+_clocksource_set_default() {
+    local validated
+    if files_load_config --optional cluster-clocksource; then
+        
+        _clocksource_validate_source $cw_CLUSTER_default_clocksource
+
+        echo $cw_CLUSTER_default_clocksource > $cw_CLOCKSOURCE_current_file
+        rm -f $cw_CLOCKSOURCE_config_file $cw_CLOCKSOURCE_sym_link
+    fi
+}
+
+_clocksource_set_source() {
+    local input="$1" validated default
+    shift
+
+    _clocksource_validate_source $input
+
+    if files_load_config --optional cluster-clocksource; then
+        default="$cw_CLUSTER_default_clocksource"
+    else
+        default=$(cat $cw_CLOCKSOURCE_current_file)
+    fi
+
+    echo $input > $cw_CLOCKSOURCE_current_file
+    _clocksource_write_config "$default" "$validated"
+    
+    if clocksource_is_master_node; then
+        ln -fs $cw_CLOCKSOURCE_config_file $cw_CLOCKSOURCE_sym_link
+    fi
+}
+
+_clocksource_ssh_compute_nodes() {
+    local master_node="$1" cmd="$2" node="$4"
+    shift 3
+
+    if [[ "$master_node" != "$node" ]]; then
+        ssh $node alces configure clocksource $cmd
+    fi
+}
+
+clocksource_is_master_node() {
+    if [[ -z "$cw_INSTANCE_role" ]]; then
+        files_load_config instance config/cluster
+    fi
+    if [[ "$cw_INSTANCE_role" == "master" ]]; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+clocksource_set() {
+    local cmd="$1"
+    shift
+
+    if [[ "$cmd" == "default" ]]; then
+        _clocksource_set_default
+    else
+        _clocksource_set_source "$cmd"
+    fi
+
+    if clocksource_is_master_node; then
+        member_each _clocksource_ssh_compute_nodes $(hostname) "$cmd"
+    fi
+}
+
+clocksource_list() {
+    local clocksources c cur indicator default_set
+    files_load_config --optional cluster-clocksource
+    clocksources="default $(_clocksource_get_available)"
+    cur=$(cat $cw_CLOCKSOURCE_current_file)
+    
+    if [[ ! -z "$cw_CLUSTER_clocksource" ]] && [[ "$cw_CLUSTER_clocksource" != "$cur" ]]; then
+        action_die "conflict between system clocksource and config file"
```

Comment:
Should this prevent the user from seeing a list of valid options? e.g. so that they can run `al conf clocksource actually-valid-option` to fix the problem?